R/estimate_uncertainty.R

```diff
@@ -0,0 +1,213 @@
+#'   pmf to compute, for each forecast date and delay, the expected number of
+#'   confirmed cases on that forecast date at that delay d. For each delay, it
+#'   estimates an independent negative binomial dispersion parameter. This code
+#'   was adapted from code written (under an MIT license) by the Karlsruhe
+#'   Institute of Technology RESPINOW German Hospitalization Nowcasting Hub,
+#' Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/39e2b17bc79492b0aee4c0b615a1c8dbf978ef53/code/baseline/functions.R#L142 #nolint
+#' @param triangle_to_nowcast matrix of the incomplete reporting triangle to be
+#'   nowcasted, with rows representing the time points of reference and columns
+#'   representing the delays
+#' @param delay_pmf vector of delays assumed to be indexed starting at the
+#'   first delay column in `triangle_to_nowcast`
+#' @param n_history_dispersion integer indicating the number of reference dates
+#'   to be used in the estimate of the dispersion, always starting from the most
+#'   recent refrence date. The default is to use the whole reporting triangle,
+#'   so `nrow(triangle_to_nowcast) - 1`
+#' @importFrom cli cli_abort
+#' @returns a vector of dispersion parameters of length of the `delay_pmf` -1
+#' @export
+#'
+#' @examples
+#' triangle <- matrix(
+#'   c(
+#'     70, 40, 20, 5,
+#'     80, 50, 10, 10,
+#'     100, 40, 31, 20,
+#'     95, 45, 21, NA,
+#'     82, 42, NA, NA,
+#'     70, NA, NA, NA
+#'   ),
+#'   nrow = 6,
+#'   byrow = TRUE
+#' )
+#' delay_pmf <- get_delay_estimate(
+#'   triangle = triangle,
+#'   max_delay = 3,
+#'   n_history = 4
+#' )
+#' disp_params <- estimate_uncertainty(
+#'   triangle_to_nowcast = triangle,
+#'   delay_pmf = delay_pmf,
+#'   n_history_dispersion = 5
+#' )
+estimate_uncertainty <- function(triangle_to_nowcast,
+                                 delay_pmf,
+                                 n_history_dispersion = nrow(triangle_to_nowcast) - 1) { # nolint
+
+  n_horizons <- length(delay_pmf) - 1
+
+  # Add validation for n_history_dispersion, must be less than
+  # number of rows in the triangle minus one
+  .validate_triangle(triangle_to_nowcast)
+  .validate_delay_and_triangle(
+    triangle = triangle_to_nowcast,
+    delay_pmf = delay_pmf
+  )
+  if (n_history_dispersion > nrow(triangle_to_nowcast) - 1) {
+    cli_abort(
+      message = c(
+        "Triangle to nowcast does not contain sufficient rows to ",
+        "estimate uncertainty from `n_history_dispersion` observations. Either",
+        "pass in a triangle of more rows or lower the `n_history_dispersion`"
+      )
+    )
+  }
+
+  # Get the truncated matrix of observations you will use to estimate the
+  # dispersion (get rid of early rows that we're not using and add NAs to
+  # bottom right of the triangle)
+  matr_observed <- triangle_to_nowcast[
+    (nrow(triangle_to_nowcast) - n_history_dispersion):
+    nrow(triangle_to_nowcast),
+  ] |>
+    .replace_lower_right_with_NA()
```

Comment:
hmm not sure I get this part?

Reply:
i.e why do it here and in the function just below

Reply:
So you don't have to do it in the `matr_observed`, however, if the `matr_observed` contains 0s rather than NAs for the bottom right missing data (related to discussion in #20 on whether we should require that missing values be NAs), then this will be treated as true observations and your expectations from `matr_observed_temp` will be compared to that. 

If we enforce that missing values should always be NAs, and 0s will be treated as observations,  then we can remove this. But as written, the other functions to estimate the delay and apply it assume that the bottom right triangle is missing.

Reply:
I can't think of another way that we would distinguish between 0s and true missingness either, which feels important.